boto3/s3/inject.py

```diff
@@ -52,7 +69,93 @@ def upload_file(self, Filename, Bucket, Key, ExtraArgs=None,
 
 def download_file(self, Bucket, Key, Filename, ExtraArgs=None,
                   Callback=None, Config=None):
+    """Download an S3 object to a file.
+
+    Usage::
+
+        import boto3
+        s3 = boto3.resource('s3')
+        s3.meta.client.download_file('mybucket', 'hello.txt', '/tmp/hello.txt')
+    """
     transfer = S3Transfer(self, Config)
     return transfer.download_file(
         bucket=Bucket, key=Key, filename=Filename,
         extra_args=ExtraArgs, callback=Callback)
+
+
+def bucket_upload_file(self, Filename, Key,
+                       ExtraArgs=None, Callback=None, Config=None):
+    """Upload a file to an S3 object.
+
+    Usage::
+
+        import boto3
+        s3 = boto3.resource('s3')
+        s3.Bucket('mybucket').upload_file('/tmp/hello.txt', 'hello.txt')
+    """
+    return self.meta.client.upload_file(
+        Filename=Filename, Bucket=self.name, Key=Key,
+        ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
+
+
+def bucket_download_file(self, Key, Filename,
+                         ExtraArgs=None, Callback=None, Config=None):
+    """Download an S3 object to a file.
+
+    Usage::
+
+        import boto3
+        s3 = boto3.resource('s3')
+        s3.Bucket('mybucket').download_file('hello.txt', '/tmp/hello.txt')
+    """
+    return self.meta.client.download_file(
+        Bucket=self.name, Key=Key, Filename=Filename,
+        ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
+
+
+def object_upload_file(self, Filename,
+                       ExtraArgs=None, Callback=None, Config=None):
+    """Upload a file to an S3 object.
+
+    Usage::
+
+        import boto3
+        s3 = boto3.resource('s3')
+        s3.Object('mybucket', 'hello.txt').upload_file('/tmp/hello.txt')
+    """
+    return self.meta.client.upload_file(
+        Filename=Filename, Bucket=self.bucket_name, Key=self.key,
+        ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
+
+
+def object_download_file(self, Filename,
+                         ExtraArgs=None, Callback=None, Config=None):
+    """Download an S3 object to a file.
+
+    Usage::
+
+        import boto3
+        s3 = boto3.resource('s3')
+        s3.Object('mybucket', 'hello.txt').download_file('/tmp/hello.txt')
+    """
+    return self.meta.client.download_file(
+        Bucket=self.bucket_name, Key=self.key, Filename=Filename,
+        ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
+
+
+functions_to_be_patched = {
+    "upload_file()": [upload_file, bucket_upload_file, object_upload_file],
+    "download_file()": [
+        download_file, bucket_download_file, object_download_file],
+}
+extra_doc = """
+
+    Similar behavior as S3Transfer's %s method,
+    except that parameters are capitalized.
+    Detailed explanations and examples can be found at S3Transfer's Usage page
+    http://boto3.readthedocs.org/en/latest/reference/customizations/s3.html#usage
+    """
+for family, funcions in functions_to_be_patched.items():
```

Comment:
We should generally avoid having code like this run in the global scope.

There's already very similar text in each of the docstrings.  I don't think it's that big of a downside if we have to repeat some documentation.

If we really want to save some copy/paste of docs, we should be doing it in a way that avoids code at the module level.

However, I'd vote to just keep it simple and add the docstrings directly.